src/settings/toml/deploy_target.rs

```diff
@@ -0,0 +1,117 @@
+use crate::settings::toml::Route;
+
+#[derive(Debug)]
+pub struct RouteConfig {
+    pub workers_dev: Option<bool>,
+    pub route: Option<String>,
+    pub routes: Option<Vec<String>>,
+    pub zone_id: Option<String>,
+}
+
+impl RouteConfig {
+    fn has_conflicting_targets(&self) -> bool {
+        if self.workers_dev.unwrap_or_default() {
+            if let Some(pattern) = &self.route {
+                !pattern.is_empty() || self.routes.is_some() // this is all so messy because of deserializer
+            } else {
+                self.routes.is_some()
+            }
+        } else {
+            if let Some(pattern) = &self.route {
+                !pattern.is_empty() && self.routes.is_some()
+            } else {
+                false
+            }
+        }
+    }
+
+    pub fn is_zoneless(&self) -> bool {
+        self.workers_dev.unwrap_or_default() && !self.has_conflicting_targets()
+    }
+
+    pub fn is_complete_zoned(&self) -> bool {
+        (self.route.is_some() || self.routes.is_some()) && self.zone_id.is_some()
+    }
+
+    // zone id is another weird one where `Some("")` is treated the same as `None`
+    pub fn missing_zone_id(&self) -> bool {
+        let result = !self.workers_dev.unwrap_or_default()
+            && (self.route.is_some() || self.routes.is_some());
+        if let Some(zone_id) = &self.zone_id {
+            result && zone_id.is_empty()
+        } else {
+            result
+        }
+    }
+}
+
+impl DeployTarget {
+    pub fn build(
+        script: &String,
```

Comment:
should accept &str not &String https://stackoverflow.com/questions/40006219/why-is-it-discouraged-to-accept-a-reference-to-a-string-string-vec-vec-o

Reply:
fixed here and in `add_route` below.